machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+type TriggerRebuild string
+
+const (
+	// RebuildNow indicates that the user wants to rebuild their image immediately
+	RebuildNow TriggerRebuild = "RebuildNow"
+	// RebuildDelayed indicates that the user wants to rebuild their image the next time the build controller is idle.
+	RebuildDelayed TriggerRebuild = "RebuildDelayed"
+	// NoRebuild indicates the user does not want a manual rebuild, this is the default.
+	NoRebuild TriggerRebuild = "NoRebuild"
+)
+
+// MachineOSRebuild describes user options to manage rebuilding the image.
+type MachineOSRebuild struct {
+	// rebuildStrategy is the user provided manual rebuild method of choice
+	// +kubebuilder:default:=NoRebuild
+	// +kubebuilder:validation:Enum:=NoRebuild;RebuildNow;RebuildDelayed
+	// +kubebuilder:validation:Required
+	RebuildStrategy TriggerRebuild `json:"rebuildStrategy"`
+	// maxRetries is the amount of times we want to auto-rebuild after build failures before we deem the update unreconcilable.
+	// +kubebuilder:default:=3
+	// +kubebuilder:validation:Required
+	MaxRetries int `json:"maxRetries"`
+}
+
+type MachineOSImageBuilderType string
+
+const (
+	// describes that the machine-os-builder will use the built in OpenshiftImageBuilder
+	OCPBuilder MachineOSImageBuilderType = "OpenShiftImageBuilder"
+	// describes that the machine-os-builder will use a custom pod builder that uses buildah
+	PodBuilder MachineOSImageBuilderType = "PodImageBuilder"
+	// Default means the cluster will use the OpenshiftImageBuilder
+	DefaultBuilder MachineOSImageBuilderType = "Default"
+)
+
+// BuildHistory contains information about related builds
+type BuildHistory struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+	// configMaps references all config map objects created during the build process
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ConfigMaps []BuildConfigMapObjectReference `json:"configmaps" patchStrategy:"merge" patchMergeKey:"name" `
+	// buildPod references the build pod used in this previous build if it existed
+	// +optional
+	BuildPod BuildPodReference `json:"buildPod"`
+}
+
+// BuildPodReference refers to the pod used to build an image that either failed or was interrupted
```

Comment:
why only failed builds?

Reply:
failed, interrupted, etc. I guess we could put in progress or succeeded builds here too :) though, we will still limit it to the most recent 3 for size constraints